apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicyElements.cs

```diff
@@ -0,0 +1,459 @@
+                _valueToObjectFunc = valueToObjectFunc ?? (value => value.ToString());
+            }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => _elementName;
+
+            /// <inheritdoc/>
+            object ISupportsExact<TValue>.ToPostPolicyValue(TValue value) =>
+                _valueToObjectFunc(value);
+
+            /// <inheritdoc/>
+            public override bool Equals(object obj) =>
+                obj is CustomElement<TValue> other ? _elementName.Equals(other._elementName) : false;
+
+            /// <inheritdoc/>
+            public override int GetHashCode() => _elementName.GetHashCode();
+        }
+
+
+        internal class CustomElementFactory : ICustomElementFactory
+        {
+            /// <summary>
+            /// The default custom element factory.
+            /// </summary>
+            public static ICustomElementFactory Default { get; } = new CustomElementFactory();
+
+            private CustomElementFactory() { }
+
+            public IPostPolicyElement Create<T>(string name, Func<T, object> valueToObject) =>
+                new CustomElement<T>(name, valueToObject);
+        }
+
+        internal sealed class PolicyAlgorithm : ISupportsExact<string>
+        {
+            /// <summary>
+            /// The default x-goog-algorithm element.
+            /// </summary>
+            public static PolicyAlgorithm Element { get; } = new PolicyAlgorithm();
+
+            internal PolicyAlgorithm() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "x-goog-algorithm";
+
+            /// <inheritdoc/>
+            object ISupportsExact<string>.ToPostPolicyValue(string value) => value;
+        }
+
+        internal sealed class PolicyCredential : ISupportsExact<string>
+        {
+            /// <summary>
+            /// The default x-goog-credential element.
+            /// </summary>
+            public static PolicyCredential Element { get; } = new PolicyCredential();
+
+            internal PolicyCredential() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "x-goog-credential";
+
+            /// <inheritdoc/>
+            object ISupportsExact<string>.ToPostPolicyValue(string value) => value;
+        }
+
+        internal sealed class PolicyCreationDateTime : ISupportsExact<DateTime>
+        {
+            /// <summary>
+            /// The default x-goog-date element.
+            /// </summary>
+            public static PolicyCreationDateTime Element { get; } = new PolicyCreationDateTime();
+
+            internal PolicyCreationDateTime() { }
+
+            /// <inheritdoc/>
+            string IPostPolicyElement.ElementName => "x-goog-date";
+
+            /// <inheritdoc/>
+            object ISupportsExact<DateTime>.ToPostPolicyValue(DateTime value) =>
+                value.ToUniversalTime().ToString("YYYYMMDD'T'HHMMSS'Z'", CultureInfo.InvariantCulture);
```

Comment:
As noted in chat, this format string won't work.

Do we definitely want to support arbitrary `DateTime` values? I'd be tempted to either use `DateTimeOffset`, or require a `DateTime` with a kind of UTC. Basically let's stop users from shooting themselves in the feet with system-time-zone-sensitive code.